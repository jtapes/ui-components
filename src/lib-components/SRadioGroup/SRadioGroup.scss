@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-radio-group {
  &__item {
    margin-bottom: $size--20;
  }

  &__items {
    @include scrollbar;

    display: flex;
    overflow-y: auto;
  }
}
