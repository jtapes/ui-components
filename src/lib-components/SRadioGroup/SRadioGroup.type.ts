import { SRadioValue, SRadioStates } from '@/lib-components/SRadio/SRadio.type'

export type SRadioGroupItem = {
  label: string
  checkedValue: SRadioValue
  state?: SRadioStates
}

export enum SRadioGroupHiddenStates {
  SHOW_MORE = 'show-more',
  SCROLL = 'scroll'
}
