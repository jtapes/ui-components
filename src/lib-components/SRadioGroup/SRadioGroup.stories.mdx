import { action } from '@storybook/addon-actions'
import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Props,
  Canvas,
} from '@storybook/addon-docs'
import SRadioGroup from '@/lib-components/SRadioGroup/SRadioGroup.vue'
import results from '/.jest-test-results.json';

<Meta
  title="DESIGN SYSTEM/Atom/RadioGroup"
  component={SRadioGroup}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    items: {
      control: 'array',
      defaultValue: [
        { label: 'Значение 1', checkedValue: 1 },
        { label: 'Значение 2', checkedValue: 2 },
        { label: 'Значение 3', checkedValue: 3 },
        { label: 'Значение 4', checkedValue: 4 },
      ]
    },
    columns: { control: 'number', defaultValue: 1 },
    searchable: { control: 'boolean', defaultValue: true },
    visibleCount: { control: 'number', defaultValue: 3 },
    hiddenState: { control: 'select', options: ['show-more', 'scroll'], defaultValue: 'show-more' }
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-Market-Catalog-%E2%80%A2-Product-Page?node-id=1902%3A3451',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# CheckboxGroup

<Description of={SRadioGroup} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SRadioGroup },
  template:
    `
      <div style="width: 400px">
        <SRadioGroup
          v-model="selectedValue"
          :items="items"
          :columns="columns"
          @change="handleChange"
          :searchable="searchable"
          :visible-count="visibleCount"
          :hidden-state="hiddenState"
        ></SRadioGroup>
      </div>
`,
  data() {
    return {
      selectedValue: null
    }
  },
  methods: {
    handleChange() {
      action('change')()
    }
  }
})

<Canvas withToolbar withSource="none">
  <Story name="SRadioGroup">
    {Template.bind({})}
  </Story>
</Canvas>

## Пример

<Source
  language="html"
  code={`
    <s-radio-group v-model="selectedValue" :items="items"/>
  `}
/>

<ArgsTable of={SRadioGroup} />
