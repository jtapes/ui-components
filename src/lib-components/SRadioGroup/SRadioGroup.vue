<template>
  <div class="s-radio-group">
    <s-input v-if="searchable" v-model="searchText">
      <template #append>
        <svg-icon name="search-24-32" class="icon-32 color-font-secondary" />
      </template>
    </s-input>
    <s-row
      v-if="filteredItems.length"
      class="s-radio-group__items"
      :class="{ 'mt-24': searchable }"
      :style="{ maxHeight: `${maxHeight}px` }"
      :gutter="20"
    >
      <s-col
        v-for="(item, index) in filteredItems"
        :key="index"
        :span="spanCount"
        class="s-radio-group__item"
      >
        <s-radio
          :checked-value="item.checkedValue"
          :label="item.label"
          :value="value"
          :state="item.state ? item.state : 'default'"
          multiple
          @change="handleValueChange"
        />
      </s-col>
    </s-row>
    <div v-else class="mt-24 color-font-secondary text-sm">
      Ничего не найдено
    </div>
    <a
      v-if="showMoreVisible"
      href="#"
      class="link d-f ai-c text-sm"
      @click.prevent="handleOpenedChange"
    >
      {{ opened ? 'Свернуть' : 'Показать еще' }}
      <svg-icon :name="opened ? 'minus-16' : 'plus-16'" class="icon-16 ml-4" />
    </a>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Model, Prop, Watch } from 'vue-property-decorator'
import SRadio from '@/lib-components/SRadio/SRadio.vue'
import SInput from '@/lib-components/SInput/SInput.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import SCol from '@/lib-components/SCol/SCol.vue'
import SRow from '@/lib-components/SRow/SRow.vue'
import type { SRadioValue } from '@/lib-components/SRadio/SRadio.type'
import type { SRadioGroupItem } from '@/lib-components/SRadioGroup/SRadioGroup.type'
import { SRadioGroupHiddenStates } from '@/lib-components/SRadioGroup/SRadioGroup.type'

const MAX_SPAN_COUNT = 24

/** Группа чекбоксов */
@Component({
  components: {
    SRadio,
    SInput,
    SvgIcon,
    SRow,
    SCol
  }
})
export default class SRadioGroup extends Vue {
  /** Массив чекбоксов */
  @Prop({ type: Array, required: true })
  items!: SRadioGroupItem[]

  /** Разбеение чекбоксов по колонкам */
  @Prop({ type: Number, default: 1 })
  columns!: number

  /** Отображение инпута для поиска */
  @Prop({ type: Boolean, default: true })
  searchable!: boolean

  /** Количество видемых чекбоксов до скрытия */
  @Prop({ type: Number, default: 3 })
  visibleCount!: number

  /** Тип скрытия */
  @Prop({
    type: String,
    default: SRadioGroupHiddenStates.SHOW_MORE,
    validator(value: string): boolean {
      return Object.values(SRadioGroupHiddenStates).includes(value as SRadioGroupHiddenStates)
    }
  })
  hiddenState!: string

  @Model('change')
  value!: SRadioValue | null

  searchText = ''
  opened = false
  maxHeight = 380

  get filteredItems(): SRadioGroupItem[] {
    let items = [...this.items]
    if (this.searchText) {
      items = items.filter((item: SRadioGroupItem) =>
        item.label.toLowerCase().includes(this.searchText.toLowerCase())
      )
    }

    if (
      items.length > this.visibleCount &&
      !this.opened &&
      this.hiddenState === SRadioGroupHiddenStates.SHOW_MORE
    ) {
      items.splice(this.visibleCount, items.length - 1)
    }

    return items
  }

  get showMoreVisible(): boolean {
    if(this.hiddenState !== SRadioGroupHiddenStates.SHOW_MORE) return false

    if (!this.searchText) {
      return this.items.length > this.visibleCount
    }

    return (
      this.items.filter((item: SRadioGroupItem) =>
        item.label.includes(this.searchText)
      ).length > this.visibleCount
    )
  }

  get spanCount(): number {
    return MAX_SPAN_COUNT / this.columns
  }

  @Watch('value')
  onValueChange(value: SRadioValue[] | null): void {
    if (!value || !value.length) {
      this.searchText = ''
    }
  }

  getContainerMaxHeight(): void {
    const items = document.querySelectorAll('.s-radio-group__item')
    let maxHeight = 0

    items.forEach((item, index) => {
      if(index + 1 > this.visibleCount) return

      const itemMargin = parseInt(window.getComputedStyle(item).marginBottom)
      maxHeight += (item.clientHeight + itemMargin)
    })

    this.maxHeight = maxHeight
  }

  handleOpenedChange(): void {
    this.opened = !this.opened
  }

  handleValueChange(value: SRadioValue[]): void {
    this.$emit('change', value)
  }

  mounted(): void {
    if(this.hiddenState === SRadioGroupHiddenStates.SCROLL) {
      this.getContainerMaxHeight()
    }
  }
}
</script>
