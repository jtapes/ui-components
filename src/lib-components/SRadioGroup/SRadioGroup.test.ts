import { mount } from '@vue/test-utils'
import SRadioGroup from '@/lib-components/SRadioGroup/SRadioGroup.vue'

describe('SRadioGroup', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SRadioGroup, {
      propsData: {
        items: [
          {
            label: '1',
            checkedValue: 1,
          },
        ],
      },
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('RadioGroup should be change visible when opened', async () => {
    const wrapper = mount(SRadioGroup, {
      propsData: {
        items: [
          {
            label: '1',
            checkedValue: 1,
          },
          {
            label: '2',
            checkedValue: 2,
          },
          {
            label: '3',
            checkedValue: 3,
          },
          {
            label: '4',
            checkedValue: 4,
          },
        ],
      },
    })

    const showMoreButton = wrapper.find('a.link')
    expect(showMoreButton.text()).toBe('Показать еще')

    let radioCount = wrapper.findAll('.s-radio-group__item').length
    expect(radioCount).toBe(3)

    await showMoreButton.trigger('click')
    await wrapper.vm.$nextTick()

    radioCount = wrapper.findAll('.s-radio-group__item').length
    expect(radioCount !== 3).toBe(true)
    expect(showMoreButton.text()).toBe('Свернуть')
  })

  test('Should be filtered by search text and emitted by $emit()', async () => {
    const wrapper = mount(SRadioGroup, {
      propsData: {
        items: [
          {
            label: '1',
            checkedValue: 1,
          },
          {
            label: '2',
            checkedValue: 2,
          },
          {
            label: '3',
            checkedValue: 3,
          },
          {
            label: '4',
            checkedValue: 4,
          },
        ],
      },
    })

    wrapper.vm.$data.searchText = '1'
    await wrapper.vm.$nextTick()

    const radioCount = wrapper.findAll('.s-radio-group__item').length
    expect(radioCount).toBe(1)

    const radio = wrapper.find('input[type="radio"]')
    await radio.setChecked()
    expect(wrapper.emitted('change')).toBeTruthy()
  })
})
