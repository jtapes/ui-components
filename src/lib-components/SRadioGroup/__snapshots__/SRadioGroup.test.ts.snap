// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SRadioGroup Is Vue snapshot 1`] = `
<div
  class="s-radio-group"
>
  <div
    class="s-input s-input--text s-input--rounded"
  >
    <!---->
     
    <input
      class="s-input__inner s-input__inner--appended"
      type="text"
    />
     
    <!---->
     
    <!---->
     
    <div
      class="s-input__append"
    >
      <svg
        class="svg-icon icon-32 color-font-secondary svg-icon--search-24-32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          href="#search-24-32"
          xlink:href="#search-24-32"
        />
      </svg>
    </div>
  </div>
   
  <div
    class="s-row s-radio-group__items s-row--justify-start s-row--align-center mt-24"
    style="margin-left: -10px; margin-right: -10px; max-height: 380px;"
  >
    <div
      class="s-col s-radio-group__item s-col-24 s-col--align-start"
      style="padding-left: 10px; padding-right: 10px;"
    >
      <label
        class="s-radio text-sm s-radio--default s-radio--default"
        multiple="multiple"
      >
        <div
          class="d-f"
        >
          <div
            class="s-radio__input"
          >
            <div
              class="s-radio__inner"
            />
             
            <input
              aria-hidden="false"
              class="s-radio__original"
              type="radio"
              value="1"
            />
          </div>
           
          <div
            class="s-radio__label"
          >
            1
          </div>
           
          <!---->
        </div>
         
        <!---->
      </label>
    </div>
  </div>
   
  <!---->
</div>
`;
